torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -347,16 +356,56 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
   /// The base case, when the list of modules is empty.
   void push_back() {}
 
-  // Box the AnyModules to give Sequential reference semantics, like the rest of
-  // the API. Note that this is not required otherwise, this could just be a
+  // Box the AnyModules to give AnySequential reference semantics, like the rest
+  // of the API. Note that this is not required otherwise, this could just be a
   // `vector<AnyModule>`.
   std::vector<AnyModule> modules_;
+
+  friend class SequentialImpl;
+};
+
+/// A `ModuleHolder` subclass for `AnySequentialImpl`.
+/// See the documentation for `AnySequentialImpl` class to learn what methods it
+/// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
+/// module storage semantics.
+TORCH_MODULE(AnySequential);
+
+/// `Sequential` is derived from `AnySequential` and has a non-template
+/// returning type allowing `Sequential`s to be nested in each other making them
+/// compatible with the python API. The template input type helps warn users
+/// about the change in the API.
+class SequentialImpl : public AnySequentialImpl {
+ public:
+  using AnySequentialImpl::AnySequentialImpl;
+
+  template <typename ReturnType = Tensor, typename... InputTypes>
+  Tensor forward(InputTypes&&... inputs) {
+    static_assert(
+        std::is_same<ReturnType, Tensor>(),
+        "Can only call Sequential::forward with Tensor as output type."
+        " If you would like to have a non-Tensor output type, please use AnySequential instead.");
+    return AnySequentialImpl::forward(std::forward<InputTypes>(inputs)...);
+  }
+
+  std::shared_ptr<Module> clone(
+      const optional<Device>& device = nullopt) const override {
+    auto clone = std::make_shared<SequentialImpl>();
+    for (const auto& module : modules_) {
+      clone->push_back(module.clone(device));
+    }
+    return std::move(clone);
+  }
```

Comment:
Is there a reason we need a special `clone` function for `SequentialImpl`?

Reply:
Because clone function of ```AnySequential``` returns ```std::make_shared<AnySequentialImpl>()``` and the ```IsClonable``` test failed before this change because we have this:
```
Sequential sequential(Linear(3, 4), Functional(torch::relu), BatchNorm(3));
  Sequential clone =
      std::dynamic_pointer_cast<SequentialImpl>(sequential->clone());
```

The error:
```
19:00:45 [ RUN      ] SequentialTest.IsCloneable
19:00:45 unknown file: error: C++ exception with description "Accessing empty ModuleHolder"
19:00:45 [  FAILED  ] SequentialTest.IsCloneable (0 ms)
```

@yf225 I may be wrong here so does this make sense to you?

Reply:
Yes I think the reason `IsClonable` test fails before this change is because we try to `dynamic_pointer_cast` the result of `sequential->clone()` from `std::shared_ptr<AnySequentialImpl>` to `std::shared_ptr<SequentialImpl>`, which returns nullptr because an `AnySequentialImpl` is not a `SequentialImpl`. 

In general, users shouldn't need to do `Sequential clone = 
 std::dynamic_pointer_cast<SequentialImpl>(sequential->clone())` and should just do `Sequential clone = sequential->clone()`, because `SequentialImpl` is an implementation detail that the user shouldn't need to care about. I think what we should do here is the following:
1. Improve the `Sequential` constructor to be able to take in a `std::shared_ptr<AnySequentialImpl>`, move all of its submodules into a new `SequentialImpl`, and then wrap the `SequentialImpl` in a new `Sequential` and return it. 
2. Remove the `clone` method in `SequentialImpl`, because it essentially has the same logic as `AnySequentialImpl`'s clone method.
3. Make sure both `Sequential clone = sequential->clone()` and `AnySequential clone = any_sequential->clone()`  works, without the need to do `dynamic_pointer_cast`. This would make our Sequential API much nicer.